Framework/Geometry/src/Objects/CSGObject.cpp

```diff
@@ -1083,20 +1083,90 @@ int CSGObject::interceptSurface(Geometry::Track &track) const {
   for (auto &surface : m_surList) {
     surface->acceptVisitor(LI);
   }
+
   const auto &IPoints(LI.getPoints());
   const auto &dPoints(LI.getDistance());
+  const auto u_vec = track.direction();
+
+  // sort the points based on its
+  // 1. build a vector that contains the relative distance to the starting point
+  //    of the track/ray
+  // 2. get the index in the sorted order
+  const size_t nPoints(IPoints.size());
+  std::vector<size_t> idxs(nPoints, 0);
+  std::vector<double> dists(nPoints);
+  if (nPoints > 1) {
+    for (size_t i = 0; i < nPoints; i++) {
+      // This will make intercept before the starting point have negative
+      // distance
+      dists[i] = u_vec.scalar_prod(IPoints[i] - track.startPoint());
+    }
+    // sort by distance, only keep index
+    std::iota(idxs.begin(), idxs.end(), 0);
+    std::sort(idxs.begin(), idxs.end(), [&dists](size_t a, size_t b) { return dists[a] < dists[b]; });
+  } else if (nPoints == 1) {
+    idxs[0] = 0;
+    dists[0] = u_vec.scalar_prod(IPoints[0] - track.startPoint());
+  }
 
-  auto ditr = dPoints.begin();
-  auto itrEnd = IPoints.end();
-  for (auto iitr = IPoints.begin(); iitr != itrEnd; ++iitr, ++ditr) {
-    if (*ditr > 0.0) // only interested in forward going points
-    {
-      // Is the point and enterance/exit Point
-      const TrackDirection flag = calcValidType(*iitr, track.direction());
-      if (flag != TrackDirection::INVALID)
-        track.addPoint(flag, *iitr, *this);
+  // 3. iterating through IPoints by the order of the smallest
+  //    relative distance to the largest
+  // - Find the first sorted point that has positive distance
+  size_t idx_firstPositive = 0;
+  for (size_t i = 0; i < nPoints; ++i) {
+    if (dists[idxs[i]] > 0) {
+      idx_firstPositive = i;
+      break;
+    }
+  }
+  // going through the list
+  for (size_t i = idx_firstPositive; i < nPoints; ++i) {
+    const size_t idx(idxs[i]);
+    // - get current point
+    const auto &currentPt(IPoints[idx]);
+    // - get previous intercept with positive distance
+    V3D prePoint;
+    if (i - 1 < idx_firstPositive) {
+      prePoint = track.startPoint();
+    } else {
+      prePoint = IPoints[idxs[i - 1]];
+    }
+    // - get next intercept with positive distance
+    V3D nextPoint;
+    if (i + 1 == nPoints) {
+      nextPoint = currentPt + (currentPt - prePoint);
```

Comment:
This assumes that the "downstream" shape is similar in thickness to upstream. I'm not sure that is a safe assumption.

Reply:
> This assumes that the "downstream" shape is similar in thickness to upstream. I'm not sure that is a safe assumption.

I think this line is only used to calculate a pseudo next point if `currentPt` is the last point. So seems OK in that case?

Reply:
For any shape that is "solid" (e.g. a cylinder, flate plate, sphere) there is only one point in the `Track`. The origin can be used as the previous point, but the origin can be (in principle) anywhere in space. 

Maybe it is best to think through how `nextPoint` can be selected based on how we tend to actually use it rather than the general case. We are always rasterizing objects and calculating lengths through them or objects that are touching them. For our simple shapes (i.e. cylindrical sample inside a cylindrical container) the ray will go from one of these two Shapes to either the source or the detector. In both of those cases the shape is the "origin" of the ray. To the source and to the detector are effectively the same since they are a direction off to infinity. Then I can think of a handful of cases for the sample in a vanadium container: